aten/src/ATen/cuda/CUDAFuture.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+
+#include <ATen/core/ivalue.h>
+#include <ATen/core/ivalue_inl.h>
+#include <ATen/core/jit_type.h>
+#include <ATen/cuda/CUDAEvent.h>
+#include <ATen/cuda/CUDAMultiStreamGuard.h>
+#include <c10/core/Allocator.h>
+#include <c10/core/Device.h>
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <c10/cuda/CUDAFunctions.h>
+#include <c10/cuda/CUDAStream.h>
+#include <c10/macros/Export.h>
+#include <c10/util/intrusive_ptr.h>
+
+namespace at { namespace cuda {
+
+struct TORCH_CUDA_API CUDAFuture : at::ivalue::Future {
+ public:
+  using at::ivalue::Future::Future;
+
+  void setDataPtrExtractor(DataPtrExtractor dataPtrExtractor) override {
+    std::unique_lock<std::mutex> lock(dataPtrExtractorMutex_);
+    dataPtrExtractor_ = std::move(dataPtrExtractor);
+  }
+
+ protected:
+  c10::intrusive_ptr<Future> createInstance(at::TypePtr type) override {
+    auto fut = c10::make_intrusive<CUDAFuture>(std::move(type));
+    // The new future needs the DataPtr extractor when it gets marked complete
+    // but this might happen immediately inline or in parallel by another
+    // thread. In both these cases this would/might happen before the user has
+    // time to set their own DataPtr extractor, which might lead to failures
+    // if the default extractor can't handle some of the user's types.
+    // Therefore we propagate our extractor.
+    fut->setDataPtrExtractor(dataPtrExtractor_);
+    return fut;
+  }
+
+  void postMarkCompletedHook(const at::IValue& value) override {
+    std::vector<bool> isCudaDeviceUsed(c10::cuda::device_count(), false);
+    for (const at::DataPtr& data_ptr : extractDataPtrs(value)) {
+      if (data_ptr.device().is_cuda()) {
+        isCudaDeviceUsed[data_ptr.device().index()] = true;
+      }
+    }
+
+    cudaEvents_ = std::make_shared<std::vector<at::cuda::CUDAEvent>>();
+    for (c10::DeviceIndex idx = 0; idx < isCudaDeviceUsed.size(); idx++) {
+      if (isCudaDeviceUsed[idx]) {
+        at::cuda::CUDAEvent cudaEvent;
+        cudaEvent.record(at::cuda::getCurrentCUDAStream(idx));
+        (*cudaEvents_).push_back(std::move(cudaEvent));
+      }
+    }
+  }
+
+  std::function<void(void)> wrapCallback(
+      std::function<void(void)> callback) override {
+    return [this, callback{std::move(callback)}]() {
```

Comment:
what's the difference of this between capture by reference? are the callbacks gonna destructed sometime but we still want them alive in the lambda?

Reply:
`std::function` AFAIK is a RAII wrapper so it might hold some state which gets destroyed once the function goes out of scope. If we captured a reference to the `std::function` that would thus risk being a dangling reference.